### PR TITLE
Add Shape::as_array()

### PR DESCRIPTION
**Pull request by odashi**
This change adds `Shape::as_array::<N>()` which returns a copy of underlying array of dimensions.
This function will be used to generate shape information of `ndarray::ArrayN`.

Blocking change: #22

### CODE REVIEW

File: src/shape.rs [hunk 1]
```diff
@@ -93,6 +93,25 @@ impl Shape {
...
+    ///
+    /// # Returns
+    ///
+    /// * `Ok([usize; N])` - Array representation of dimensions.
+    /// * `Err(Error)` - The shape is not N-dimensional.
+    pub fn as_array<const N: usize>(&self) -> Result<[usize; N]> {
+        if N == self.num_dimensions {
+            let mut v = MaybeUninit::<[usize; N]>::uninit();
+            unsafe { &mut *v.as_mut_ptr() }.copy_from_slice(&self.dimensions[..N]);
+            Ok(unsafe { v.assume_init() })
```
Comment (vbkaisetsu) [inline src/shape.rs]: https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.as_mut_ptr
> Reading from this pointer or turning it into a reference is undefined behavior unless the MaybeUninit<T> is initialized.

However, the result of `as_mut_ptr()` is turned into a reference before initialization in this code.
The current code may work, but why not use the implementation of `uninit_array()`, `write_slice()`, and `array_assume_init()` provided by Nightly Rust?

```suggestion
            // let mut v: [MaybeUninit<usize>; N] = MaybeUninit::uninit_array();
            let mut v = unsafe { MaybeUninit::<[MaybeUninit<usize>; N]>::uninit().assume_init() };
            // MaybeUninit::write_slice(&mut v, &self.dimensions[..N]);
            let uninit_dimensions: &[MaybeUninit<T>] = unsafe { std::mem::transmute(self.dimensions[..N]) };
            v.copy_from_slice(uninit_dimensions);
            // Ok(unsafe { MaybeUninit::array_assume_init(v) })
            Ok(unsafe { (&v as *const _ as *const [usize; N]).read() })
```

Comment (odashi) [reply src/shape.rs]: because they are nightly.

Comment (odashi) [reply src/shape.rs]: I think this UB would become serious sometimes and needs to be avoided if possible.

Comment (vbkaisetsu) [reply src/shape.rs]: What does UB mean?

Comment (vbkaisetsu) [reply src/shape.rs]: ahh, undefined behaviour

Comment (odashi) [reply src/shape.rs]: It seems the current version of the language could not define types related to const parameters. This implementation needs to be modified.

Comment (odashi) [reply src/shape.rs]: fixed in 928d375, now we have separate definitions for all `as_array[0-8]`.

Comment (odashi) [reply src/shape.rs]: I guessed this is why ndarray defines separate ArrayN instead of using generics...

**Review by vbkaisetsu — APPROVED** [review]

**Comment by odashi** [comment]
@tetsuok ping

**Review by tetsuok — APPROVED** [review]

**Pull request merged**